test/pylib/scylla_cluster.py

```diff
@@ -6,6 +6,7 @@
 """Scylla clusters for testing.
```

Comment:
This should be three separate commits:
- test/pylib: scylla_cluster: introduce ScyllaServer start-stop lock
- test/pylib: scylla_cluster: introduce task history
- test/pylib: scylla_cluster: fix teardown race with ongoing test operation

(and in last commit, explain how the lock and the history are used to fix the race)

Reply:
Could you please describe why?

because this 2 logic can not work separately 

with start-stop lock without tasks cancelling/awaiting - we will still faced with issues teardown
and vice versa is the same

Reply:
You don't need to solve all problems in a single commit.

You can use one commit to introduce some concept which will be useful later to solve a problem, then another commit to introduce yet another concept, then another commit that uses the previously introduced concepts to actually solve some problem.

Generally we try to split commits in scylladb repo on module or feature boundaries. If you introduce two modules or two features, they should be in separate commits.

This helps reviewers keep track of what's happening.

Otherwise when I review one large commit and read it top to bottom I see:
- fragment of code related to feature/concept X
- then fragment of code related to feature/concept Y
- then fragment related to X again

etc. and it's hard for me to understand everything because I have limited memory in my head.

In your commit message you even have two headers:
"Introducing the Start-Stop Lock:"
and
"Introducing Tasks History:"

so these two concepts can definitely be understood independently of each other.

Then you connect two separate concepts so they work together to solve a problem.

But they are still separate concepts. So they should be introduced in separate commits.

Reply:
split in 2. not sure what code change needs to be added in 3rd commit